### PR TITLE
feat: rewrite prometheus analytics to clickhouse

### PR DESCRIPTION
**Pull request by Pagebakers**
## Motivation and Context

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts [hunk 1]
```diff
@@ -178,215 +127,392 @@ export class MetricsDashboardRepository {
...
+        previousValue: parseValue(prevMedianResponse[0]?.medianRate),
+        top: top5Response.map((v) => ({
+          name: v.name,
+          value: parseValue(v.value),
+        })),
+        series: this.mapSeries(range, seriesResponse, prevSeriesResponse),
       },
-      errors: responses.filter((r) => r.error),
-    };
+    } as any;
```
Comment (StarpTech) [inline controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:218]: Please create proper types.

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts [hunk 2]
```diff
@@ -178,215 +127,392 @@ export class MetricsDashboardRepository {
...
+        SELECT
+          Timestamp as timestamp,
+        OperationName as name,
+          sum(TotalRequests) as rate
+        FROM operation_request_metrics_5_30_mv
+        WHERE Timestamp >= startDate AND Timestamp <= endDate
+          AND OrganizationID = '${params.organizationId}'
+          AND FederatedGraphID = '${params.graphId}'
+        GROUP BY Timestamp, OperationName 
+      ) GROUP BY name ORDER BY value DESC LIMIT 5
```
Comment (StarpTech) [inline controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:172]: Let's show the top 10.

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts [hunk 3]
```diff
@@ -178,215 +127,392 @@ export class MetricsDashboardRepository {
...
+
+    const [start, end] = getDateRange(endDate, range);
+    const [prevStart, prevEnd] = getDateRange(endDate, range, range);
+
+    const queryLatency = (quantile: string, start: number, end: number) => {
+      return this.chClient.queryPromise<{ value: number }>(`
+        WITH
+          toDateTime('${start}') AS startDate,
+          toDateTime('${end}') AS endDate
+        SELECT quantile(0.5)(value) as value FROM (
```
Comment (StarpTech) [inline controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:235]: This looks incorrect. Building a media on top of the PX quantile is odd. We can just return the computed quantile `func_histogram_v2` for all operations.

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts [hunk 4]
```diff
@@ -178,215 +127,392 @@ export class MetricsDashboardRepository {
...
-      query: `topk(5, sum by (wg_operation_name) (rate(cosmo_router_http_request_duration_milliseconds_bucket{${this.getQueryLabels(
-        params,
-      )}}[${range}h])))`,
-    });
+    const queryTop5 = (quantile: string, start: number, end: number) => {
+      return this.chClient.queryPromise<{ name: string; value: string }>(`
+      WITH
+        toDateTime('${start}') AS startDate,
+        toDateTime('${end}') AS endDate
+      SELECT name, round(quantileDeterministic(0.5)(value, 1), 4) AS value FROM (
```
Comment (StarpTech) [inline controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:273]: Same here. We can rank the operation by P95.

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts [hunk 5]
```diff
@@ -178,215 +127,392 @@ export class MetricsDashboardRepository {
...
+    // get median request rate in last [range]h
+    const value = this.chClient.queryPromise<{ errorPercentage: number }>(`
+      WITH
+        toDateTime('${start}') AS startDate,
+        toDateTime('${end}') AS endDate
+      SELECT
+        sum(totalErrors) AS serverErrors,
+        sum(totalClientErrors) as clientErrors,
+        sum(totalRequests) AS requests,
+        round(serverErrors + clientErrors, 3) as errors,
```
Comment (StarpTech) [inline controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:387]: totalErrors already includes client errors.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts (hunk 1 above)
Comment (Pagebakers) [reply controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts]: Done, left overs from debugging

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts (hunk 3 above)
Comment (Pagebakers) [reply controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:228]: I wasn't getting correct value until I added this, this gets the median value of the p95 in the time range to my understanding.

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts (hunk 4 above)
Comment (Pagebakers) [reply controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:266]: Same as above, get the median value of the aggregated time range

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts (hunk 5 above)
Comment (Pagebakers) [reply controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts]: Updated

File: controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts (hunk 2 above)
Comment (Pagebakers) [reply controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts:165]: Sticking with 5 in this view, more will occupy too much space for other charts that will be added here.

I'm adding a `view all` link under each metric to the traces page grouped by operation name and sorted by requests/latency/error rate

**Pull request merged**